src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs

```diff
@@ -71,19 +89,194 @@ protected override void EndProcessing()
             {
                 return;
             }
-            orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
 
-            if (_unique)
+            // If -Unique was used, if -Top & -Bottom were not used, or of -Top or -Bottom would return all objects, sort
+            // all objects, remove duplicates if necessary, and identify the list of items to return
+            if (_unique || (Top == 0 && Bottom == 0) || Top >= orderByProperty.OrderMatrix.Count || Bottom >= orderByProperty.OrderMatrix.Count)
             {
-                RemoveDuplicates(orderByProperty);
+                orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
+
+                if (_unique)
+                {
+                    RemoveDuplicates(orderByProperty);
+                }
+
+                var listToOutput = orderByProperty.OrderMatrix;
+
+                if (Top > 0 && Top < listToOutput.Count)
+                {
+                    // There may be a faster algorithm to sort while eliminating duplicates and retaining only
+                    // the top N records; for now though, this gets the job done
+                    listToOutput = listToOutput.GetRange(0, Top);
+                }
+                else if (Bottom > 0 && Bottom < listToOutput.Count)
+                {
+                    int lastItemIndex = listToOutput.Count - 1;
+                    listToOutput = listToOutput.GetRange(lastItemIndex - Bottom, lastItemIndex);
+                }
+
+                // write to output stream
+                foreach (var x in listToOutput)
+                {
+                    WriteObject(x.inputObject);
+                }
+            }
+            // Otherwise, use an indexed min-/max-heap to sort all objects
+            else
+            {
+                // The heap entry comparer will sort the items in the heap
+                var comparer = new IndexedMinMaxHeapComparer(orderByProperty.Comparer);
+
+                // Identify how many items will be in the heap
+                int indexedMinMaxHeapSize = Top > 0 ? Top : Bottom;
+
+                // A small list will store the data for the heap 
+                var indexedMinMaxHeap = new List<IndexedMinMaxHeapEntry>(indexedMinMaxHeapSize + 1);
```

Comment:
@iSazonov It's probably a misunderstanding on my part as well. I've put a lot of effort into this PR, and I'm trying to wrap it up so that I can move on to other work. Your last comment sounded like you were still doubting that the heap was required, after I've implemented and tested it with it sorting data more than twice as fast with Top N/Bottom N sorts. That's why I reacted the way I did, but I understand your comments are meant to be constructive and impartial. The performance improvements I'm seeing in my latest revision are fantastic (including for -unique, which I made work much, much faster, e.g. 5 times faster for large data sets, with and without Top N/Bottom N). I'm looking forward to this commit, which will be done soon (just need to clean up the Pester test cases as per the feedback I have received).